### ISSUE TITLE
Mercure hub not working in production while working locally

### ISSUE THREAD

**Issue by Marko1212** [dunglas/mercure#668]
Hi,

I'm not able to use the Mercure HUB in production, whereas everything works fine on local machine.

The error I get, in production, is : 

Access to resource at 'http://localhost:3000/.well-known/mercure?topic=http%3A%2F%2Fmysite.local%2Fwebnotifications%2F1' from origin 'https://dev.mysite.fr' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

This is a Symfony application with Docker, and I am using docker-compose.

Mercure configuration : 

  appli_mercure:
    image: dunglas/mercure
    container_name: appli_mercure
    restart: unless-stopped
    networks:
      - core
    ports:
      - "3000:3000"
    environment:
      SERVER_NAME: ":3000"
      MERCURE_PUBLISHER_JWT_KEY: 'a_very_long_string'
      MERCURE_SUBSCRIBER_JWT_KEY: 'a_very_long_string'
      # Set the URL of your Symfony project (without trailing slash!) as value of the cors_origins directive
      MERCURE_EXTRA_DIRECTIVES: |
        cors_origins http://mysite.local
    # Comment the following line to disable the development mode
    command: /usr/bin/caddy run -config /etc/caddy/Caddyfile.dev
    volumes:
      - mercure_data:/data
      - mercure_config:/config

My JS file : 

...
const url = new URL('http://localhost:3000/.well-known/mercure');
url.searchParams.append('topic', `an_url_here`);
eventSource = new EventSource(url);
...

So I tried to add anonymous in cors_origins, in docker-compose.yml file, but it's still not working.

Clearly, locally the app is under http, while in production it is under https.

Is it possible to keep Mercure running under http, both locally and in production (http://localhost:3000/.well-known/mercure), and make it work both with the app running under http locally (http://mysite.local, already works !) and under https in production ( https://dev.mysite.fr, this is NOT working) ?

Thanks in advance!

**Comment by dunglas** [comment]
This looks like a duplicate of https://github.com/dunglas/mercure/issues/618.
Can you double-check your configuration?

**Comment by algorun77** [comment]
Hello,

I'm working on this project with @Marko1212, and I'd like to add additional information.

We are implementing Mercure using the [Symfony tutorial](https://symfony.com/doc/5.4/mercure.html), and we honestly don't know much of either Mercure or Caddy or even Docker as we're not the one who set up the project. So, our issue probably reside in our current configuration.

We first have our `docker-compose.yml` file:
```yml
  dj_mercure:
    image: dunglas/mercure
    container_name: dj_mercure
    restart: unless-stopped
    ports:
      - "3000:3000"
    networks:
      - core
    environment:
      SERVER_NAME: ':3000'
      MERCURE_PUBLISHER_JWT_KEY: '${MERCURE_SECRET}'
      MERCURE_SUBSCRIBER_JWT_KEY: '${MERCURE_SECRET}'
      # Set the URL of your Symfony project (without trailing slash!) as value of the cors_origins directive
      MERCURE_EXTRA_DIRECTIVES: |
        cors_origins http://${HOST}
    # Comment the following line to disable the development mode
    #command: /usr/bin/caddy run -config /etc/caddy/Caddyfile.dev
    volumes:
      - mercure_data:/data
      - mercure_config:/config

networks:
  core:
    driver: bridge

volumes:
  mercure_data:
  mercure_config:
```

Here, the MERCUE_SECRET and HOST variables are set in the Docker .env file. For instance, let's say our server URL is : https://dev.site.com, then `HOST=dev.site.com`.

Then, we have our project directly. Publishing with Symfony seems to work fine as we didn't come across any error. Our template has the following lines:
```javascript
const url = new URL('http://localhost:3000/.well-known/mercure');
url.searchParams.append('topic', `http://thefastjourney.local/webnotifications/${connectedUserId}`);
eventSource = new EventSource(url);
```

In our local project, which run in HTTP, everything works fine at that point. But not on the server (https://dev.site.com) where we are triggering a CORS policy error. Example with Firefox :
Blocked loading mixed active content "http://dev.thefastjourney.eu/_wdt/c4de3f" notifications:298:3366
Uncaught Error: no element is specified to initialize PerfectScrollbar
    H https://dev.thefastjourney.eu/lib/amaretti-2.0.0/js/app.js:804
    initScroller https://dev.thefastjourney.eu/lib/amaretti-2.0.0/js/app.js:70
    dropdownsScrollerInit https://dev.thefastjourney.eu/lib/amaretti-2.0.0/js/app.js:553
    init https://dev.thefastjourney.eu/lib/amaretti-2.0.0/js/app.js:699
    <anonymous> https://dev.thefastjourney.eu/fr/admin/notifications:163
    jQuery 13
app.js:804:13789

`Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://localhost:3000/.well-known/mercure?topic=xyz. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing). Status code: 200.`

`Firefox can’t establish a connection to the server at http://localhost:3000/.well-known/mercure?topic=xyz. notifications:106:30
Error: Promised response from onMessage listener went out of scope 6 background.js:841:170`

It seems that whatever the value of cors_orgins we use, we can't make this work.

We looked through a lot of similar issues but nothing works. I'm a little suspicious because everyone seems to have a Caddy file but we don't. When printing the content of the `/etc/caddy/Caddyfile` **inside the container**, this is what we have:
```
# Learn how to configure the Mercure.rocks Hub on https://mercure.rocks/docs/hub/config
{
	{$GLOBAL_OPTIONS}
}

{$SERVER_NAME:localhost}

log

route {
	encode zstd gzip

	mercure {
		# Transport to use (default to Bolt)
		transport_url {$MERCURE_TRANSPORT_URL:bolt://mercure.db}
		# Publisher JWT key
		publisher_jwt {env.MERCURE_PUBLISHER_JWT_KEY} {env.MERCURE_PUBLISHER_JWT_ALG}
		# Subscriber JWT key
		subscriber_jwt {env.MERCURE_SUBSCRIBER_JWT_KEY} {env.MERCURE_SUBSCRIBER_JWT_ALG}
		# Extra directives
		{$MERCURE_EXTRA_DIRECTIVES}
	}

	respond /healthz 200

	respond "Not Found" 404
}
```

**Where is the cors_origins line from the docker-compose file?**

Our Docker contains also a Nginx server, and this is the configuration we have right now:
```
server {
    listen 80 default_server;

    server_name localhost;
    root /var/www/html/public;
    index index.php index.html index.htm;

    location / {
        # try to serve file directly, fallback to index.php
        try_files $uri /index.php$is_args$args;
    }

    # Dev
    location ~ ^/(index_dev|config)\.php(/|$) {
        fastcgi_pass dj_php:9000;
        fastcgi_split_path_info ^(.+\.php)(/.*)$;
        include fastcgi_params;
        fastcgi_param SCRIPT_FILENAME $realpath_root$fastcgi_script_name;
        fastcgi_param DOCUMENT_ROOT $realpath_root;
    }

    # Prod
    location ~ ^/index\.php(/|$) {
        fastcgi_pass dj_php:9000;
        fastcgi_split_path_info ^(.+\.php)(/.*)$;
        include fastcgi_params;
        fastcgi_param SCRIPT_FILENAME $realpath_root$fastcgi_script_name;
        fastcgi_param DOCUMENT_ROOT $realpath_root;
        internal;
    }

    # return 404 for all other php files not matching the front controller
    # this prevents access to other php files you don't want to be accessible.
    location ~ \.php$ {
        return 404;
    }

   # return 404 for all other php files not matching the front controller
   # this prevents access to other php files you don't want to be accessible.
   location ~ \.php$ {
     return 404;
   }

}
```

Is there anything we missed or did wrong? Should we have a configuration similar to the one from the issue you mentioned?

Thanks in advance!

**Comment by algorun77** [comment]
Well,

In the end, we managed to make it work. This is what we added to our Nginx default.conf file :
```
location /.well-known/mercure {
    proxy_pass http://dj_mercure:3000;
    proxy_read_timeout 24h;
    proxy_http_version 1.1;
    proxy_set_header Connection "";

    ## Be sure to set USE_FORWARDED_HEADERS=1 to allow the hub to use those headers ##
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Host $host;
    proxy_set_header X-Forwarded-Proto $scheme;
}
```

Our issue was mainly that we used `http://127.0.0.1:3000` as Mercure hub URL bu this localhost point to the Nginx container, not the main Host or Mercure. Since our containers share the network, this works.

We also updated our Javascript file with this :
```javascript
const url = new URL('https://dev.site.com:3000/.well-known/mercure');
```
Instead of localhost, which worked fine in our local machines.

And finally, the docker-compose file :
```yaml
  dj_mercure:
    environment:
      MERCURE_EXTRA_DIRECTIVES: |
        cors_origins http://${HOST}
        use_forwarded_headers "1"
```
With that, it seems to be finally working!
